pkg/reconciler/ingress/config/gateway.go

```
@@ -0,0 +1,148 @@
+	// labels matching a particular selector, it will use the
+	// corresponding gateway.  If multiple selectors match, we choose
+	// the most specific selector.
+	Gateways map[v1alpha1.IngressVisibility]*GatewayConfig
+}
+
+// NewGatewayFromConfigMap creates a Gateway from the supplied ConfigMap
+func NewGatewayFromConfigMap(configMap *corev1.ConfigMap) (*Gateway, error) {
+	v, ok := configMap.Data[visibilityConfigKey]
+	if !ok {
+		// These are the defaults.
+		return &Gateway{
+			Gateways: map[v1alpha1.IngressVisibility]*GatewayConfig{
+				v1alpha1.IngressVisibilityExternalIP:   {GatewayClass: defaultGatewayClass, Gateway: defaultIstioGateway, Service: defaultGatewayService},
+				v1alpha1.IngressVisibilityClusterLocal: {GatewayClass: defaultGatewayClass, Gateway: defaultIstioLocalGateway, Service: defaultLocalGatewayService},
+			},
+		}, nil
+	}
+
+	entry := make(map[v1alpha1.IngressVisibility]GatewayConfig)
+	if err := yaml.Unmarshal([]byte(v), &entry); err != nil {
+		return nil, err
+	}
+
+	for _, vis := range []v1alpha1.IngressVisibility{
+		v1alpha1.IngressVisibilityClusterLocal,
+		v1alpha1.IngressVisibilityExternalIP,
+	} {
+		if _, ok := entry[vis]; !ok {
+			return nil, fmt.Errorf("visibility %q must not be empty", vis)
+		}
+	}
+	c := Gateway{Gateways: map[v1alpha1.IngressVisibility]*GatewayConfig{}}
+
+	for key, value := range entry {
+		key, value := key, value
+		// Check that the visibility makes sense.
+		switch key {
+		case v1alpha1.IngressVisibilityClusterLocal, v1alpha1.IngressVisibilityExternalIP:
+		default:
+			return nil, fmt.Errorf("unrecognized visibility: %q", key)
+		}
+
+		// See if the Service is a valid namespace/name token.
+		if _, _, err := cache.SplitMetaNamespaceKey(value.Service); err != nil {
+			return nil, err
+		}
+
+		// See if the Gateway is a valid namespace/name token.
+		if _, _, err := cache.SplitMetaNamespaceKey(value.Gateway); err != nil {
+			return nil, err
+		}
```

From the comment at the using side: Should we store `types.NamespacedName` in `Gateways` type so that calling sides can use this without having to ignore errors and in a structured way already?